lib/stm32/fdcan.c

```diff
@@ -0,0 +1,891 @@
+
+	if (nr >= rxgfc_lss_current)
+	{
+		return FDCAN_E_OUTOFRANGE;
+	}
+
+	ram->LFSSA[nr].type_id1_conf_id2 = 
+		((id_list_mode << FDCAN_SFT_SHIFT) & FDCAN_SFT_MASK)
+		| (action & FDCAN_SFEC_MASK)
+		| ((id1 << FDCAN_SFID1_SHIFT) & FDCAN_SFID1_SHIFT)
+		| ((id2 << FDCAN_SFID2_SHIFT) & FDCAN_SFID2_SHIFT);
+
+	return 0;
+}
+
+/** Configure filter rule for extended ID frames.
+ * Sets up filter rule for frames having extended ID. Each FDCAN block can
+ * have its own set of filtering rules. It is only possible to configure as
+ * many filters as was configured previously using @see fdcan_filter_init.
+ * @param canport FDCAN block base address
+ * @param nr number of filter to be configured
+ * @param id_list_mode mode in which id1 and id2 are used to match the rule, see 
+ * FDCAN_EFT_<mode> defines.
+ * @param id1 extended ID for matching. Used as exact value, lower bound or bit 
+ * pattern depending on matching mode selected
+ * @param id2 extended ID or bitmask. Used as exact value, upper bound or bit mask 
+ * depending on matching mode selected
+ * @param action Action performed if filtering rule matches frame ID, see 
+ * FDCAN_EFEC_<action> defines.
+ * @returns FDCAN_E_OK if filter was set up properly, 
+ * FDCAN_E_INVALID if FDCAN block base address is unknown and 
+ * FDCAN_E_OUTOFRANGE is filter number is larger than previosly configured.
+ */
+int fdcan_ext_filter_set(uint32_t canport, uint32_t nr, 
+             uint8_t id_list_mode, uint32_t id1, uint32_t id2,
+		     uint8_t action)
+{
+	struct FDCAN_MessageRAM * ram = fdcan_get_msgram_addr(canport);
+
+	if (!ram)
+	{
+		return FDCAN_E_INVALID;
+	}
+
+	uint32_t rxgfc_lse_current = ((FDCAN_RXGFC(canport) & FDCAN_RXGFC_LSE_MASK) 
+			>> FDCAN_RXGFC_LSE_SHIFT);
+
+	if (nr >= rxgfc_lse_current)
+	{
+		return FDCAN_E_OUTOFRANGE;
+	}
+
+	ram->LFESA[nr].conf_id1 = 
+		(action & FDCAN_EFEC_MASK)
+		| ((id1 << FDCAN_EFID1_SHIFT) & FDCAN_EFID1_SHIFT);
+
+	ram->LFESA[nr].type_id2 = 
+		(id_list_mode & FDCAN_EFT_MASK)
+		| ((id2 << FDCAN_EFID2_SHIFT) & FDCAN_EFID2_SHIFT);
+
+	return 0;
+}
+
+
+/** @brief CAN Transmit Message
+ *
+ * @param canport Unsigned int32. CAN block register base @ref can_reg_base.
+ * @param id Unsigned int32. Message ID.
+ * @param ext bool. Extended message ID?
+ * @param rtr bool. Request transmit?
+ * @param fdcan_fmt Use FDCAN format?
+ * @param btr_switch Switch bitrate for data portion of frame?
+ * @param length Unsigned int8. Message payload length.
+ * @param data Unsigned int8[]. Message payload data.
+ * @returns int 0, 1 or 2 on success and depending on which outgoing mailbox got
+ * selected. On error returns FDCAN_E_BUSY if no mailbox was available,
+ * FDCAN_E_INVALID if CAN block base address and/or message length are invalid.
+ */
+int fdcan_transmit(uint32_t canport, uint32_t id, bool ext, bool rtr,
+        bool fdcan_fmt, bool btr_switch, uint8_t length, const uint8_t *data)
+{
+	int mailbox;
+	uint32_t dlc, flags = 0;
```

Comment:
personally, I don't see any benefit to this sort of misra style, but I'm not going to complain about it.   (... much ;)